src/script/interpreter.cpp

```diff
@@ -459,6 +459,47 @@ static bool EvalChecksigPreTapscript(const valtype& vchSig, const valtype& vchPu
     return true;
 }
 
+static bool EvalTapScriptCheckSigFromStack(const valtype& sig, const XOnlyPubKey& pubkey, ScriptExecutionData& execdata, unsigned int flags, const Span<const unsigned char> msg, SigVersion sigversion, ScriptError* serror, bool& success)
+{
+    // This code follows the behaviour of EvalCheckSigTapscript
+    assert(sigversion == SigVersion::TAPSCRIPT);
+
+    /*
+     *  The following validation sequence is consensus critical. Please note how --
+     *    upgradable public key versions precede other rules;
+     *    the script execution fails when using empty signature with invalid public key;
+     *    the script execution fails when using non-empty invalid signature.
+     */
+    success = !sig.empty();
+    if (success) {
+        // Implement the sigops/witnesssize ratio test.
+        // Passing with an upgradable public key version is also counted.
+        assert(execdata.m_validation_weight_left_init);
+        execdata.m_validation_weight_left -= VALIDATION_WEIGHT_PER_SIGOP_PASSED;
+        if (execdata.m_validation_weight_left < 0) {
+            return set_error(serror, SCRIPT_ERR_TAPSCRIPT_VALIDATION_WEIGHT);
+        }
+    }
+    if (pubkey.size() == 0) {
+        return set_error(serror, SCRIPT_ERR_PUBKEYTYPE);
+    } else if (pubkey.size() == 32) {
+        if (success && !pubkey.VerifySchnorr(msg, sig)) {
+            return false; // serror is set
```

Comment:
No `serror` has not been set.

I think test coverage must be lacking here.

Reply:
yes, it is. I have only asserted the failure of the test in the test suite and not the actual error code for it. I will update the tests so that they assert the real reason why they fail.

Reply:
Ah right. This error is less severe than I had thought.

Reply:
In any case, I am updating the test suite to check for actual return error codes.